src/docfx/restore/index/DependencyGitPool.cs

```diff
@@ -0,0 +1,175 @@
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+
+            if (!_acquiredGits.TryGetValue((key, gitVersion.Commit, LockType.Shared), out var gitInfo))
+            {
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+
+            if (string.IsNullOrEmpty(gitInfo.path) || gitInfo.git == null)
+            {
+                throw Errors.NeedRestore($"{remote}#{branch}").ToException();
+            }
+
+            var path = Path.Combine(AppData.GetGitDir(remote), gitInfo.path);
+            Debug.Assert(Directory.Exists(path));
+
+            return (path, gitVersion);
+        }
+
+        /// <summary>
+        /// Try get git dependency repository path and git slot with remote, branch and dependency version(commit).
+        /// If the dependency version is null, get the latest one(order by last write time).
+        /// If the dependency version is not null, get the one matched with the version(commit).
+        /// </summary>
+        public static async Task<(string path, DependencyGit git)> TryGetGit(string remote, string branch, string commit)
+        {
+            var (path, slot) = await DependencySlotPool.TryGetSlot<DependencyGit>(remote, gits =>
+            {
+                var filteredGits = gits.Where(i => i.Branch == branch);
+                if (!string.IsNullOrEmpty(commit))
+                {
+                    // found commit matched slot
+                    filteredGits = filteredGits.Where(i => i.Commit == commit);
+                }
+
+                return filteredGits.ToList();
+            });
+
+            if (!string.IsNullOrEmpty(path))
+                path = Path.Combine(AppData.GetGitDir(remote), path);
+
+            return !Directory.Exists(path) ? default : (path, slot);
+        }
+
+        public static async Task<(string path, DependencyGit git)> AcquireExclusiveGit(string remote, string branch, string commit)
+        {
+            var (path, git) = await AcquireGit(remote, branch, commit, LockType.Exclusive);
+
+            Debug.Assert(path != null && git != null);
+            path = Path.Combine(AppData.GetGitDir(remote), path);
+
+            return (path, git);
+        }
+
+        private static Task<(string path, DependencyGit git)> AcquireGit(string remote, string branch, string commit, LockType type)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(branch));
+            Debug.Assert(!string.IsNullOrEmpty(commit));
+
+            return DependencySlotPool.AcquireSlots<DependencyGit>(
+                remote,
+                type,
+                slot =>
+                {
+                    // update branch and commit info to new rented slot
+                    slot.Commit = commit;
+                    slot.Branch = branch;
+                    return slot;
+                },
+                existingSlot => existingSlot.Branch == branch && existingSlot.Commit == commit);
```

Comment:
`commit` alone is enough to identify a slot, since we are not looking up `branch` from repo.

Reply:
not enough, we have `@implicit` feature which based on branch.